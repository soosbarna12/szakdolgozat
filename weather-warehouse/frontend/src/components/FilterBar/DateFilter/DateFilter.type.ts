import { Dayjs } from "dayjs";

export interface DateFilterProps {
  onDateChange?: (dateValue: Dayjs | null) => void;
}