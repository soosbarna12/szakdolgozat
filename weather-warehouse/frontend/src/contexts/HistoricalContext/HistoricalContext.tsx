import React, { createContext, useEffect, useState } from "react";
import { LOCAL_STORAGE_TEMPERATURE_SCALE } from "../../consts/temperatureScale.const";
import { TemperatureScale } from "../../types/temperatureScale.type";
import { HistoricalContextType, HistoricalLocationData, HistoricalPageData, HistoricalProviderProps } from "./HistoricalContext.type";


export const HistoricalContext = createContext<HistoricalContextType>({
  location: { name: "", lat: 0, lon: 0 },
  setLocation: () => { },
  historicalPageData: [] as any,
  setHistoricalPageData: () => { },
  temperatureScale: TemperatureScale.Celsius,
  setTemperatureScale: () => { },
});

export const HistoricalLocationProvider = ({ children }: HistoricalProviderProps) => {

  const [location, setLocation] = useState<HistoricalLocationData>(() => {
    const storedLocation = localStorage.getItem("historicalLocation");
    const defaultLocation = { name: "", lat: 0, lon: 0 };
    try {
      return storedLocation ? JSON.parse(storedLocation) : defaultLocation;
    } catch (error) {
      console.error("Failed to parse location from localStorage:", error);
      return defaultLocation;
    }
  });

  const [historicalPageData, setHistoricalPageData] = useState<HistoricalPageData[]>(() => {
    const storedData = localStorage.getItem("historicalPageData");
    try {
      return storedData ? JSON.parse(storedData) : [];
    } catch (error) {
      console.error("Failed to parse historical data from localStorage:", error);
      return [];
    }
  });

  const [temperatureScale, setTemperatureScale] = useState<TemperatureScale | string>(() => {
    const storedTemperatureScale = localStorage.getItem(LOCAL_STORAGE_TEMPERATURE_SCALE);
    try {
      return storedTemperatureScale ? JSON.parse(storedTemperatureScale) : TemperatureScale.Celsius;
    } catch (error) {
      console.error("Failed to parse temperature scale from localStorage:", error);
      return TemperatureScale.Celsius;
    }
  });


  useEffect(() => {
    if (historicalPageData) {
      localStorage.setItem("historicalPageData", JSON.stringify(historicalPageData));
    }
  }, [historicalPageData]);

  useEffect(() => {
    localStorage.setItem("historicalLocation", JSON.stringify(location));
  }, [location]);

  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_TEMPERATURE_SCALE, JSON.stringify(temperatureScale));
  }, [temperatureScale]);


  return (
    <HistoricalContext.Provider value={{ location, setLocation, historicalPageData, setHistoricalPageData, temperatureScale, setTemperatureScale }}>
      {children}
    </HistoricalContext.Provider>
  );
};
