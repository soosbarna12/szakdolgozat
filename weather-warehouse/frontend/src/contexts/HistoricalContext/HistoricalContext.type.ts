import { Dispatch, ReactNode, SetStateAction } from "react";
import { HistoricalDataTable } from "../../types/historicalDataTable.type";
import { TemperatureScale } from "../../types/temperatureScale.type";

export interface HistoricalLocationData {
  name: string;
  country?: string;
  lat: number;
  lon: number;
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface HistoricalPageData extends HistoricalDataTable {}

export interface HistoricalContextType {
  location: HistoricalLocationData;
  historicalPageData: any [];
  temperatureScale: TemperatureScale | string;
  setLocation: Dispatch<SetStateAction<HistoricalLocationData>>;
  setHistoricalPageData: Dispatch<SetStateAction<HistoricalPageData[]>>;
  setTemperatureScale: Dispatch<SetStateAction<TemperatureScale | string>>;
}

export interface HistoricalProviderProps {
  children: ReactNode;
}
