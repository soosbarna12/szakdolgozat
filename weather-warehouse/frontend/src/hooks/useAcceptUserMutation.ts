import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useAcceptUserMutation() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (id: number) => {
      const response = await axios.post(`/user/accept`, { id });
      return response.data;
    },
    onSuccess: () => {
      showAlert("User accepted successfully", "success");
    },
    onError: () => {
      showAlert("Failed to accept user", "error");
    },
  });

  return mutation;
}
