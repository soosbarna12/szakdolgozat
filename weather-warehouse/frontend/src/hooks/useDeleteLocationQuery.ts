import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";

export function useDeleteLocationQuery() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (userLocationID: number) => {
      const response = await axios.delete(`/user/deleteLocation`, { params: { userLocationID } });
      return response.data;
    },
    onSuccess: () => {
      showAlert("Location deleted successfully", "success");
    },
    onError: () => {
      showAlert("Failed to delete location", "error");
    },
  });

  return mutation;
}