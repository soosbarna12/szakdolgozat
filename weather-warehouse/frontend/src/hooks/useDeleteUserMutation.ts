import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useDeleteUserMutation() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (id: number) => {
      const response = await axios.delete(`/user/delete`, { params: { id } });
      return response.data;
    },
    onSuccess: () => {
      showAlert('User deleted successfully', 'success');
    },
    onError: () => {
      showAlert('Failed to delete user', 'error');
    },
  });

  return mutation;
}
