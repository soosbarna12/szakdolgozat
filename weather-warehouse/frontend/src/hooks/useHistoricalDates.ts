import { useQuery } from "@tanstack/react-query";
import dayjs from "dayjs";
import { useEffect } from "react";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";

export function useHistoricalDates(location: string) {
  const { showAlert } = useAlert();

  const { data, error, isLoading } = useQuery({
    queryKey: ["historicalDates", location],
    queryFn: async () => {
      const response = await axios.get(`/api/historical/historicalDates`, {
        params: { location },
      });
      
      return response.data.map((date: any) => dayjs(date.date).format("YYYY-MM-DD"));
    },
    
    enabled: !!(location),
    refetchOnWindowFocus: false,
  });

  useEffect(() => {
    if (error) {
      showAlert("Error fetching historical dates", "error");
    }
  }, [error]);

  return { data: data || [], error, isLoading };
}
