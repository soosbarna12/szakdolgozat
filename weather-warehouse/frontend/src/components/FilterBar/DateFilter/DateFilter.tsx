import { LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import React, { useContext, useEffect, useRef } from "react";
import { StyledDatePicker } from "../../../stlyes/inputField.style";
import { DateFilterProps } from "./DateFilter.type";
import { useState } from "react";
import dayjs, { Dayjs } from "dayjs";
import { HistoricalContext } from "../../../contexts/HistoricalContext/HistoricalContext";
import { useHistoricalDates } from "../../../hooks/useHistoricalDates";
import { LOCAL_STORAGE_SELECTED_DATE_NAME } from "../../../consts/historicalDate.const";


export function DateFilter({ onDateChange }: DateFilterProps) {

  const [selectedDate, setSelectedDate] = useState<Dayjs | null>(getStoredDate());
  const { location } = useContext(HistoricalContext);
  const { data: historicalDates, isLoading } = useHistoricalDates(location.name);
  const isInitialized = useRef(false);
  const previousLocation = useRef(location.name);

  function getStoredDate() {
    const storedDate = localStorage.getItem(LOCAL_STORAGE_SELECTED_DATE_NAME);
    return storedDate ? dayjs(storedDate) : null
  }

  const handleDateChange = (date: Dayjs | null) => {
    setSelectedDate(date);
    onDateChange?.(date);
  };

  // reset datepicker when location changes
  useEffect(() => {
    if (isInitialized.current && location.name !== previousLocation.current) {
      setSelectedDate(null);
      onDateChange?.(null);
    } else {
      isInitialized.current = true;
    }
    previousLocation.current = location.name;
  }, [location.name]);

  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_SELECTED_DATE_NAME, selectedDate?.format("YYYY-MM-DD") ?? "");
  }, [selectedDate]);

  return (
    <LocalizationProvider dateAdapter={AdapterDayjs}>
      <StyledDatePicker
        format="YYYY-MM-DD"
        sx={{ boxShadow: 4 }}
        onAccept={handleDateChange}
        value={selectedDate}
        minDate={dayjs("1750-01-01")} // Set the minimum date to the first date in historicalDates
        disabled={!location.name || isLoading || !historicalDates?.length} // disable if no location is selected
        shouldDisableDate={(date) => {
          if (isLoading || !historicalDates?.length) return true; // Disable all dates while loading
          const formattedDate = date.format("YYYY-MM-DD");
          return !historicalDates.includes(formattedDate); // Disable dates not in historicalDates
        }}
        slotProps={{
          textField: {
            size: 'small',
            placeholder: selectedDate
              ? selectedDate.format("YYYY-MM-DD")
              : "Date",
          },
          popper: { sx: { ".MuiPaper-root": { borderRadius: "20px" } } },
        }}
      />
    </LocalizationProvider>
  );
}
