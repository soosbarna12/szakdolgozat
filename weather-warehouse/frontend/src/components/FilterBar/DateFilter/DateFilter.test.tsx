import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import dayjs from 'dayjs';
import React from 'react';
import { HistoricalContext } from '../../../contexts/HistoricalContext/HistoricalContext';
import { useHistoricalDates } from '../../../hooks/useHistoricalDates';
import { renderWithQueryClient } from '../../../utils/test/renderWithQueryClient';
import { DateFilter } from './DateFilter';

jest.mock('../../../hooks/useHistoricalDates', () => ({
    useHistoricalDates: jest.fn(),
}));

describe('FilterBar/DateFilter', () => {
    const mockOnDateChange = jest.fn();
    const mockHistoricalDates = ['2023-01-01', '2023-01-02', '2023-01-03'];

    beforeEach(() => {
        jest.clearAllMocks();
        (useHistoricalDates as jest.Mock).mockImplementation(() => ({
            data: mockHistoricalDates,
            isLoading: false,
        }));

        Object.defineProperty(window, 'localStorage', {
            value: {
                getItem: jest.fn((key) => {
                    if (key === 'selectedDate') {
                        return null;
                    }
                    return null;
                }),
                setItem: jest.fn(),
            },
            writable: true,
        });
    });

    const renderComponent = (locationName = 'Test Location', selectedDate = null) => {
        window.localStorage.getItem = jest.fn(() =>
            selectedDate ? dayjs(selectedDate).format('YYYY-MM-DD') : null
        );

        return renderWithQueryClient(
            <HistoricalContext.Provider value={{ location: { name: locationName } as any } as any}>
                <DateFilter onDateChange={mockOnDateChange} />
            </HistoricalContext.Provider>
        );
    };

    it('matches DateFilter component snapshot', () => {
        const component = renderComponent();
        expect(component).toMatchSnapshot();
    });

    it('renders the date picker with the correct placeholder when no date is selected', () => {
        renderComponent();
        const datePicker = screen.getByPlaceholderText('Date');
        expect(datePicker).toHaveAttribute('placeholder', 'Date');
    });

    it('renders the date picker with the selected date', () => {
        renderComponent('Test Location', '2023-01-01');
        const datePicker = screen.getByPlaceholderText('2023-01-01');
        expect(datePicker).toHaveValue('2023-01-01');
    });

    it('disables the date picker when no location is selected', () => {
        renderComponent('');
        const datePicker = screen.getByPlaceholderText('Date');
        expect(datePicker).toBeDisabled();
    });

    it('disables dates not in historicalDates', () => {
        renderComponent();
        const datePicker = screen.getByPlaceholderText('Date');
        const enabledDate = dayjs('2023-01-01');

        expect(datePicker).toBeInTheDocument();
        expect(
            (useHistoricalDates as jest.Mock).mock.calls[0][0]
        ).toBe('Test Location');
        expect(mockHistoricalDates).toContain(enabledDate.format('YYYY-MM-DD'));
    });
});