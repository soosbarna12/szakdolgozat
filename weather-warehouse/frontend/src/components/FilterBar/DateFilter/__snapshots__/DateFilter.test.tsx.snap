// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar/DateFilter matches DateFilter component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root css-qkn3r8-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-13meb6w-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r0»"
            inputmode="text"
            placeholder="Date"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-elo8k2-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1ysp02-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiFormControl-root MuiTextField-root css-qkn3r8-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-13meb6w-MuiInputBase-input-MuiOutlinedInput-input"
          id="«r0»"
          inputmode="text"
          placeholder="Date"
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-elo8k2-MuiInputAdornment-root"
        >
          <button
            aria-label="Choose date"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1ysp02-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="CalendarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
            </svg>
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-w4cd9x"
          >
            <span
              aria-hidden="true"
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
